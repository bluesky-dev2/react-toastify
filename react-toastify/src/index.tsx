export { useToastContainer } from './hooks';
export { cssTransition } from './utils';
export { ToastContainer, Bounce, Flip, Slide, Zoom } from './components';
export { toast } from './core';
export * from "./types"