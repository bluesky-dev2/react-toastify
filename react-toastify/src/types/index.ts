import * as React from 'react';

export type ToastPosition =
  | 'top-right'
  | 'top-center'
  | 'top-left'
  | 'bottom-right'
  | 'bottom-center'
  | 'bottom-left';

export type TypeOptions = 'info' | 'success' | 'warning' | 'error' | 'default';
export type ToastContent = React.ReactNode | (() => React.ReactNode );
export interface Toast {
  content: ToastContent;
  options: InternalToastOptions;
}

type Id = number | string
export type ToastId = Id;
export type ContainerId = Id;

interface CommonOptions {
  /**
   * Pause the timer when the mouse hover the toast.
   * `Default: true`
   */
  pauseOnHover?: boolean;

  /**
   * Pause the toast when the window loose focus.
   * `Default: true`
   */
  pauseOnFocusLoss?: boolean;

  /**
   * Remove the toast when clicked.
   * `Default: true`
   */
  closeOnClick?: boolean;

  /**
   * Set the delay in ms to close the toast automatically.
   * Use `false` to prevent the toast from closing.
   * `Default: 5000`
   */
  autoClose?: number | false;

  /**
   * Set the default position to use.
   * `One of: 'top-right', 'top-center', 'top-left', 'bottom-right', 'bottom-center', 'bottom-left'`
   * `Default: 'top-right'`
   */
  position?: ToastPosition;

  /**
   * Pass a custom close button.
   * To remove the close button pass `false`
   */
  closeButton?: React.ReactNode | boolean;

  /**
   * An optional css class to set for the progress bar.
   */
  progressClassName?: string | object | null;

  /**
   * An optional style to set for the progress bar.
   */
  progressStyle?: object;

  /**
   * An optional css class to set.
   */
  className?: string | object | null;

  /**
   * An optional css class to set for the toast content.
   */
  bodyClassName?: string | object | null;

  /**
   * Hide or show the progress bar.
   * `Default: false`
   */
  hideProgressBar?: boolean;

  /**
   * Pass a custom transition built with react-transition-group.
   */
  transition?: React.ComponentType;

  /**
   * Allow toast to be draggable
   * `Default: true`
   */
  draggable?: boolean;

  /**
   * The percentage of the toast's width it takes for a drag to dismiss a toast
   * `Default: 80`
   */
  draggablePercent?: number;

  /**
   * Define the ARIA role for the toast
   * `Default: alert`
   *  https://www.w3.org/WAI/PF/aria/roles
   */
  role?: string;

  /**
   * Set id to handle multiple container
   */
  containerId?: ContainerId;

  /**
   * Fired when clicking inside toaster
   */
  onClick?: Function;

  /**
   * Support right to left display.
   * `Default: false`
   */
  rtl?: boolean;
}

export interface ToastOptions extends CommonOptions {
  /**
   * Called when toast is mounted.
   */
  onOpen?: () => void;

  /**
   * Called when toast is unmounted.
   */
  onClose?: () => void;

  /**
   * Set the toast type.
   * `One of: 'info', 'success', 'warning', 'error', 'default'`
   */
  type?: TypeOptions;

  /**
   * Set a custom `toastId`
   */
  toastId?: ToastId;

  /**
   * Used during update
   */
  updateId?: ToastId;

  /**
   * Set the percentage for the controlled progress bar. `Value must be between 0 and 1.`
   */
  progress?: number | string;

  /**
   * Add a delay in ms before the toast appear.
   */
  delay?: number;
}

export interface InternalToastOptions extends ToastOptions {
  staleId?: ToastId;
  toastId: ToastId;
  key: Id;
  closeToast: () => void;
  position: ToastPosition;
}

export interface UpdateOptions extends ToastOptions {
  /**
   * Used to update a toast.
   * Pass any valid ReactNode(string, number, component)
   */
  render?: ToastContent;
}

export interface ToastContainerProps extends CommonOptions {
  /**
   * Whether or not to display the newest toast on top.
   * `Default: false`
   */
  newestOnTop?: boolean;

  /**
   * An optional inline style to apply.
   */
  style?: object;

  /**
   * An optional css class for the toast.
   */
  toastClassName?: string | object | null;

  /**
   * Show the toast only if it includes containerId and it's the same as containerId
   * `Default: false`
   */
  enableMultiContainer?: boolean;
}

// interface Toast {
//   /**
//    * Shorthand to display toast of type 'success'.
//    */
//   success(content: ToastContent, options?: ToastOptions): ToastId;

//   /**
//    * Shorthand to display toast of type 'info'.
//    */
//   info(content: ToastContent, options?: ToastOptions): ToastId;

//   /**
//    * Shorthand to display toast of type 'warning'.
//    */
//   warn(content: ToastContent, options?: ToastOptions): ToastId;

//   /**
//    * Shorthand to display toast of type 'error'.
//    */
//   error(content: ToastContent, options?: ToastOptions): ToastId;

//   /**
//    * Check if a toast is active by passing the `toastId`.
//    * Each time you display a toast you receive a `toastId`.
//    */
//   isActive(toastId: ToastId): boolean;

//   /**
//    * Remove a toast. If no `toastId` is used, all the active toast
//    * will be removed.
//    */
//   dismiss(toastId?: ToastId): void;

//   /**
//    * Update an existing toast. By default, we keep the initial content and options of the toast.
//    */
//   update(toastId: ToastId, options?: UpdateOptions): void;

//   /**
//    * Listen for change when a toast is added or removed. The number of toast displayed is passed as paran to the callback
//    */
//   onChange(
//     callback: (count?: number, containerId?: string | number) => void
//   ): void;

//   /**
//    * Set a controlled progress bar value to 100% then close the toast
//    */
//   done(toastId: ToastId): void;

//   /**
//    * Let you define `ToastContainer` props when lazy mounted.
//    * When called enable lazy mounted container
//    */
//   configure(config?: ToastContainerProps): void;

//   /**
//    * Display a toast without a specific type.
//    */
//   (content: ToastContent, options?: ToastOptions): ToastId;

//   /**
//    * Helper to set notification type
//    */
//   TYPE: Type;

//   /**
//    * Helper to set position
//    */
//   POSITION: Position;
// }

// /**
//  * Helper to set notification type
//  */

// /**
//  * Helper to set position
//  */

// /**
//  * Helper to build custom entrance and exit transition
//  */

// /**
//  * Built-in entrance and exit transition
//  */

// /**
//  * Built-in entrance and exit transition
//  */

// /**
//  * Built-in entrance and exit transition
//  */

// /**
//  * Built-in entrance and exit transition
//  */
