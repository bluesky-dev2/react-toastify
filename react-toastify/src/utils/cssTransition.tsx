import React from 'react';
import { Transition } from 'react-transition-group';
import { TransitionProps } from 'react-transition-group/Transition';

import { NOOP } from './constant';

export interface CSSTransitionProps {
  /**
   * Css class to apply when toast enter
   */
  enter: string;

  /**
   * Css class to apply when toast leave
   */
  exit: string;

  /**
   * Define the duration of the transition in ms
   * You can also pass an array `[enterDuration, exitDuration]`
   * `Default: 750`
   */
  duration?: number | [number, number];

  /**
   * Append current toast position to the classname.
   * For instance `myclass--top-center`...
   * `Default: false`
   */
  appendPosition?: boolean;
}

interface AnimationProps extends Partial<TransitionProps> {
  position?: Position;
  preventExitTransition?: boolean;
  
}

export function cssTransition({
  enter,
  exit,
  duration = 750,
  appendPosition = false
}: CSSTransitionProps) {
  const Animation: React.FC<AnimationProps> = ({
    children,
    position,
    preventExitTransition,
    ...props
  }) => {
    const enterClassName = appendPosition ? `${enter}--${position}` : enter;
    const exitClassName = appendPosition ? `${exit}--${position}` : exit;
    let enterDuration: number, exitDuration: number;

    if (Array.isArray(duration) && duration.length === 2) {
      [enterDuration, exitDuration] = duration;
    } else {
      enterDuration = exitDuration = duration as number;
    }

    const onEnter = (node: HTMLElement) => {
      node.classList.add(enterClassName);
      node.style.animationFillMode = 'forwards';
      node.style.animationDuration = `${enterDuration * 0.001}s`;
    };
    const onEntered = (node: HTMLElement) => {
      node.classList.remove(enterClassName);
      node.style.cssText = '';
    };
    const onExit = (node: HTMLElement) => {
      node.classList.add(exitClassName);
      node.style.animationFillMode = 'forwards';
      node.style.animationDuration = `${exitDuration * 0.001}s`;
    };

    return (
      <Transition
        {...props}
        timeout={
          preventExitTransition
            ? 0
            : {
                enter: enterDuration,
                exit: exitDuration
              }
        }
        onEnter={onEnter}
        onEntered={onEntered}
        onExit={preventExitTransition ? NOOP : onExit}
      >
        {children}
      </Transition>
    );
  };
  return Animation;
}
