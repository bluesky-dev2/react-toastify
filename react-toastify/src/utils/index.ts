export * from './propValidator';
export * from './eventManager';
export * from './constant';
export * from './cssTransition';
