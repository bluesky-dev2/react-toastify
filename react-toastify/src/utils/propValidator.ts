import { isValidElement } from 'react';

export function isValidDelay<T>(val: T): boolean {
  return typeof val === 'number' && !isNaN(val) && val > 0;
}

export function objectValues<T>(obj: Record<string, T>) {
  return Object.keys(obj).map(key => obj[key]);
}

export const canUseDom = !!(
  typeof window !== 'undefined' &&
  window.document &&
  window.document.createElement
);

// function withRequired(fn) {
//   if (process.env.NODE_ENV === 'development') {
//     fn.isRequired = function(props, propName, componentName) {
//       const prop = props[propName];

//       if (typeof prop === 'undefined') {
//         return new Error(`The prop ${propName} is marked as required in
//         ${componentName}, but its value is undefined.`);
//       }

//       fn(props, propName, componentName);
//     };
//   }
//   return fn;
// }

// export const falseOrDelay = withRequired((props, propName, componentName) => {
//   const prop = props[propName];

//   if (prop !== false && !isValidDelay(prop)) {
//     return new Error(`${componentName} expect ${propName}
//       to be a valid Number > 0 or equal to false. ${prop} given.`);
//   }

//   return null;
// });

export function canBeRendered<T>(content: T): boolean {
  return (
    isValidElement(content) ||
    typeof content === 'string' ||
    typeof content === 'number' ||
    typeof content === 'function'
  );
}

export function parseClassName(prop: string | object) {
  if (typeof prop === 'string') {
    return prop;
  } else if (prop !== null && typeof prop === 'object' && 'toString' in prop) {
    return prop.toString();
  }

  return null;
}
