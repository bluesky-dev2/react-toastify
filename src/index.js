import ToastContainer from './ToastContainer';
import toastify from './toastify';

export { ToastContainer, toastify as toast };
