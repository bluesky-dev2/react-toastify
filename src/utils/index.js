export * from './constant';
export * from './propValidator';
export * from './eventManager';
export * from './cssTransition';
