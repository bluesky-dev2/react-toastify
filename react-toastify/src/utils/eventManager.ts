import { InternalToastOptions, ToastId, ToastContent } from '../types';

export type Event = 'show' | 'clear' | 'didMount' | 'willUnmount' | 'change';

type OnShowCallback = (content: ToastContent , options: InternalToastOptions) => void;
type OnClearCallback = (id?: ToastId) => void;
type OnDidMountCallback = (containerInstance: object) => void;
type OnWillUnmountCallback = OnDidMountCallback;
type OnChangeCallback = (toast: number, containerId?: number | string) => void;
type Callback =
  | OnShowCallback
  | OnClearCallback
  | OnDidMountCallback
  | OnWillUnmountCallback
  | OnChangeCallback;
type TimeoutId = ReturnType<typeof window.setTimeout>;

export interface EventManager {
  list: Map<Event, Callback[]>;
  emitQueue: Map<Event, TimeoutId[]>;
  on(event: 'show', callback: OnShowCallback): EventManager;
  on(event: 'clear', callback: OnClearCallback): EventManager;
  on(event: 'didMount', callback: OnDidMountCallback): EventManager;
  on(event: 'willUnmount', callback: OnWillUnmountCallback): EventManager;
  on(event: 'change', callback: OnChangeCallback): EventManager;
  off(event: Event): EventManager;
  cancelEmit(event: Event): EventManager;
  emit(event: 'show', content: React.ReactNode, options: InternalToastOptions): void;
  emit(event: 'clear', id?: string | number): void;
  emit(event: 'didMount', containerInstance: object): void;
  emit(event: 'willUnmount', containerInstance: object): void;
  emit(event: 'change', toast: number, containerId?: number | string): void;
}

export const eventManager: EventManager = {
  list: new Map(),
  emitQueue: new Map(),

  on(event: Event, callback: Callback) {
    this.list.has(event) || this.list.set(event, []);
    this.list.get(event)!.push(callback);
    return this;
  },

  off(event) {
    this.list.delete(event);
    return this;
  },

  cancelEmit(event) {
    const timers = this.emitQueue.get(event);
    if (timers) {
      timers.forEach((timer: TimeoutId) => clearTimeout(timer));
      this.emitQueue.delete(event);
    }

    return this;
  },

  /**
   * Enqueue the event at the end of the call stack
   * Doing so let the user call toast as follow:
   * toast('1')
   * toast('2')
   * toast('3')
   * Without setTimemout the code above will not work
   */
  emit(event: Event, ...args: any[]) {
    this.list.has(event) &&
      this.list.get(event)!.forEach((callback: Callback) => {
        const timer = setTimeout(() => {
          // @ts-ignore
          callback(...args);
        }, 0);

        this.emitQueue.has(event) || this.emitQueue.set(event, []);
        this.emitQueue.get(event)!.push(timer);
      });
  }
};
