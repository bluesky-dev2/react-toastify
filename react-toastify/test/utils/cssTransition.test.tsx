import React from 'react';
import { render } from '@testing-library/react';

import '../__mocks__/react-transition-group';
import { cssTransition } from '../../src/utils';

describe('cssTransition helper', () => {
  it('Should return a valid react component', () => {
    const Transition = cssTransition({
      enter: 'foo',
      exit: 'bar'
    });

    const Child = () => <div>Plop</div>;
    const { container, getByText } = render(
      <Transition>
        <Child />
      </Transition>
    );

    expect(getByText('Plop').textContent).toContain('Plop');
    expect(container).toMatchSnapshot();
  });
});
