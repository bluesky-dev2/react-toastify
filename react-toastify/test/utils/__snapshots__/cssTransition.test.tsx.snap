// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cssTransition helper Should return a valid react component 1`] = `
<div>
  <div>
    <div>
      Plop
    </div>
  </div>
</div>
`;
